sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2241,6 +2240,18 @@ class Dataset[T] private[sql](
     Offset(Literal(n), logicalPlan)
   }
 
+  // This breaks caching, but it's usually ok because it addresses a very specific use case:
+  // using union to union many files or partitions.
+  private def combineUnions(u: Union): LogicalPlan = {
+    val newChildren = u.children.flatMap {
+      case child: Union if u.byName == child.byName && u.allowMissingCol == child.allowMissingCol =>
+        child.children
+      case other =>
+        Seq(other)
+    }
+    u.withNewChildren(newChildren)
```

Comment:
`withNewChildren` needs to ensure the size of children is consistent.  Can we use `Union(newChildren)`?